**Importing important libraries for the data cleaning**

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

Loading the data set

In [ ]:
df=pd.read_csv("/content/mobile.csv")

In [ ]:
df

Name  Spec Score  rating   price  \
0                            OPPO Reno 14 Pro 5G          89    4.65   41990   
1                                OPPO Reno 14 5G          87    4.75   32990   
2                                     Poco F7 5G          83    4.75   31999   
3                                   Vivo X200 FE          89    4.65   49990   
4                                   OPPO K13x 5G          73    4.20   11999   
...                                          ...         ...     ...     ...   
1014              Realme GT 7 Pro Racing Edition          85    4.05   36999   
1015                                iQOO Neo 11R          75    4.45   29990   
1016  Samsung Galaxy S25 Plus (12GB RAM + 512GB)          88    4.25  111999   
1017    Samsung Galaxy A14 4G (6GB RAM + 128 GB)          74    4.00    9699   
1018        Tecno Spark 30C 5G (8GB RAM + 128GB)          75    4.65   12999   

                                                    img       tag  \
0     https://cdn1.smartprix.com/rx-is822PXo3-w280-h...  UPCOMING   
1     https://cdn1.smartprix.com/rx-iRGgfcGDH-w280-h...  UPCOMING   
2     https://cdn1.smartprix.com/rx-icmgBU9Q2-w280-h...  LAUNCHED   
3     https://cdn1.smartprix.com/rx-iHI7IaQgQ-w280-h...  UPCOMING   
4     https://cdn1.smartprix.com/rx-iXUulomIY-w280-h...  LAUNCHED   
...                                                 ...       ...   
1014  https://cdn1.smartprix.com/rx-iA2dbnJIw-w280-h...  UPCOMING   
1015  https://cdn1.smartprix.com/rx-iBdtueAix-w280-h...   RUMORED   
1016  https://cdn1.smartprix.com/rx-iGH33GWY9-w280-h...  LAUNCHED   
1017  https://cdn1.smartprix.com/rx-igcSDy9DV-w280-h...  LAUNCHED   
1018  https://cdn1.smartprix.com/rx-ilU0Yr0Ho-w280-h...  LAUNCHED   

                                                    sim  \
0     Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
1     Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
2     Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
3     Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
4                    Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi   
...                                                 ...   
1014  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
1015  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
1016      Dual Sim, 3G, 4G, 5G, VoLTE, Vo5G, Wi-Fi, NFC   
1017                     Dual Sim, 3G, 4G, VoLTE, Wi-Fi   
1018  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   

                                              processor  \
0         Dimensity 8450, Octa Core, 3.25 GHz Processor   
1         Dimensity 8350, Octa Core, 3.35 GHz Processor   
2      Snapdragon 8s Gen4, Octa Core, 3.2 GHz Processor   
3     Dimensity 9300 Plus, Octa Core, 3.25 GHz Proce...   
4          Dimensity 6300, Octa Core, 2.4 GHz Processor   
...                                                 ...   
1014  Snapdragon 8 Elite, Octa Core, 4.32 GHz Processor   
1015                                Octa Core Processor   
1016   Snapdragon 8Elite, Octa Core, 4.47 GHz Processor   
1017             Exynos 850, Octa Core, 2 GHz Processor   
1018       Dimensity 6300, Octa Core, 2.4 GHz Processor   

                        storage                                   battery  \
0     12 GB RAM, 256 GB inbuilt   6200 mAh Battery with 80W Fast Charging   
1      8 GB RAM, 256 GB inbuilt   6000 mAh Battery with 80W Fast Charging   
2     12 GB RAM, 256 GB inbuilt   7550 mAh Battery with 90W Fast Charging   
3     12 GB RAM, 256 GB inbuilt   6500 mAh Battery with 90W Fast Charging   
4      4 GB RAM, 128 GB inbuilt   6000 mAh Battery with 45W Fast Charging   
...                         ...                                       ...   
1014  12 GB RAM, 256 GB inbuilt  6500 mAh Battery with 120W Fast Charging   
1015  12 GB RAM, 256 GB inbuilt  6700 mAh Battery with 100W Fast Charging   
1016  12 GB RAM, 512 GB inbuilt   4900 mAh Battery with 45W Fast Charging   
1017   6 GB RAM, 128 GB inbuilt   5000 mAh Battery with 15W Fast C

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1019 entries, 0 to 1018
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Name            1019 non-null   object 
 1   Spec Score      1019 non-null   int64  
 2   rating          1019 non-null   float64
 3   price           1019 non-null   int64  
 4   img             1019 non-null   object 
 5   tag             1019 non-null   object 
 6   sim             1019 non-null   object 
 7   processor       993 non-null    object 
 8   storage         999 non-null    object 
 9   battery         1013 non-null   object 
 10  display         1018 non-null   object 
 11  camera          1008 non-null   object 
 12  memoryExternal  668 non-null    object 
 13  version         971 non-null    object 
 14  fm              294 non-null    object 
dtypes: float64(1), int64(2), object(12)
memory usage: 119.5+ KB


In [ ]:
df.describe()

Spec Score       rating          price
count  1019.000000  1019.000000    1019.000000
mean     78.513248     4.377380   35006.714426
std      12.754081     0.234293   40671.714570
min       7.000000     3.450000      99.000000
25%      75.000000     4.150000   13999.000000
50%      81.000000     4.400000   23498.000000
75%      85.000000     4.600000   39990.000000
max      96.000000     4.750000  489990.000000

In [ ]:
df.isna().sum()

Name                0
Spec Score          0
rating              0
price               0
img                 0
tag                 0
sim                 0
processor          26
storage            20
battery             6
display             1
camera             11
memoryExternal    351
version            48
fm                725
dtype: int64

Missing values

processor (26 missing)

storage (20 missing)

battery (6 missing)

display (1 missing)

camera (11 missing)

memoryExternal (351 missing)

version (48 missing)

fm (725 missing → most of the column is empty)
You’ll need to decide whether to impute (fill with mode/NA/“Unknown”) or drop columns with too many missing values (like fm).

In [ ]:
#for to fill out the null values to the column we are using simple imputer
#because all columns which contain the null values are same data type

from sklearn.impute import SimpleImputer

df.select_dtypes("object").columns

Index(['Name', 'img', 'tag', 'sim', 'processor', 'storage', 'battery',
       'display', 'camera', 'memoryExternal', 'version', 'fm'],
      dtype='object')

In [ ]:
si=SimpleImputer(strategy="most_frequent")
n=si.fit_transform(df[['Name', 'img', 'tag', 'sim', 'processor', 'storage', 'battery',
       'display', 'camera', 'memoryExternal', 'version', 'fm']])

In [ ]:
a=pd.DataFrame(n,columns=df.select_dtypes("object").columns)

In [ ]:
df.drop(columns=("Name"),inplace=True)

In [ ]:
df["Name"]=a["Name"]

In [ ]:
df.drop(columns=['img', 'tag', 'sim', 'processor', 'storage', 'battery',
       'display', 'camera', 'memoryExternal', 'version', 'fm'],inplace=True)

In [ ]:
df[[ 'img', 'tag', 'sim', 'processor', 'storage', 'battery',
       'display', 'camera', 'memoryExternal', 'version', 'fm']]=a[[ 'img', 'tag', 'sim', 'processor', 'storage', 'battery',
       'display', 'camera', 'memoryExternal', 'version', 'fm']]

In [ ]:
df

Spec Score  rating   price                                        Name  \
0             89    4.65   41990                         OPPO Reno 14 Pro 5G   
1             87    4.75   32990                             OPPO Reno 14 5G   
2             83    4.75   31999                                  Poco F7 5G   
3             89    4.65   49990                                Vivo X200 FE   
4             73    4.20   11999                                OPPO K13x 5G   
...          ...     ...     ...                                         ...   
1014          85    4.05   36999              Realme GT 7 Pro Racing Edition   
1015          75    4.45   29990                                iQOO Neo 11R   
1016          88    4.25  111999  Samsung Galaxy S25 Plus (12GB RAM + 512GB)   
1017          74    4.00    9699    Samsung Galaxy A14 4G (6GB RAM + 128 GB)   
1018          75    4.65   12999        Tecno Spark 30C 5G (8GB RAM + 128GB)   

                                                    img       tag  \
0     https://cdn1.smartprix.com/rx-is822PXo3-w280-h...  UPCOMING   
1     https://cdn1.smartprix.com/rx-iRGgfcGDH-w280-h...  UPCOMING   
2     https://cdn1.smartprix.com/rx-icmgBU9Q2-w280-h...  LAUNCHED   
3     https://cdn1.smartprix.com/rx-iHI7IaQgQ-w280-h...  UPCOMING   
4     https://cdn1.smartprix.com/rx-iXUulomIY-w280-h...  LAUNCHED   
...                                                 ...       ...   
1014  https://cdn1.smartprix.com/rx-iA2dbnJIw-w280-h...  UPCOMING   
1015  https://cdn1.smartprix.com/rx-iBdtueAix-w280-h...   RUMORED   
1016  https://cdn1.smartprix.com/rx-iGH33GWY9-w280-h...  LAUNCHED   
1017  https://cdn1.smartprix.com/rx-igcSDy9DV-w280-h...  LAUNCHED   
1018  https://cdn1.smartprix.com/rx-ilU0Yr0Ho-w280-h...  LAUNCHED   

                                                    sim  \
0     Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
1     Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
2     Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
3     Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
4                    Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi   
...                                                 ...   
1014  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
1015  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
1016      Dual Sim, 3G, 4G, 5G, VoLTE, Vo5G, Wi-Fi, NFC   
1017                     Dual Sim, 3G, 4G, VoLTE, Wi-Fi   
1018  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   

                                              processor  \
0         Dimensity 8450, Octa Core, 3.25 GHz Processor   
1         Dimensity 8350, Octa Core, 3.35 GHz Processor   
2      Snapdragon 8s Gen4, Octa Core, 3.2 GHz Processor   
3     Dimensity 9300 Plus, Octa Core, 3.25 GHz Proce...   
4          Dimensity 6300, Octa Core, 2.4 GHz Processor   
...                                                 ...   
1014  Snapdragon 8 Elite, Octa Core, 4.32 GHz Processor   
1015                                Octa Core Processor   
1016   Snapdragon 8Elite, Octa Core, 4.47 GHz Processor   
1017             Exynos 850, Octa Core, 2 GHz Processor   
1018       Dimensity 6300, Octa Core, 2.4 GHz Processor   

                        storage                                   battery  \
0     12 GB RAM, 256 GB inbuilt   6200 mAh Battery with 80W Fast Charging   
1      8 GB RAM, 256 GB inbuilt   6000 mAh Battery with 80W Fast Charging   
2     12 GB RAM, 256 GB inbuilt   7550 mAh Battery with 90W Fast Charging   
3     12 GB RAM, 256 GB inbuilt   6500 mAh Battery with 90W Fast Charging   
4      4 GB RAM, 128 GB inbuilt   6000 mAh Battery with 45W Fast Charging   
...                         ...                                       ...   
1014  12 GB RAM, 256 GB inbuilt  6500 mAh Battery with 120W Fast Charging   
1015  12 GB RAM, 256 GB inbuilt  6700 mAh Battery with 100W Fast Charging   
1016  12 GB RAM, 512 GB inbuilt   4900 mAh Battery with 45W Fast Charging   
1017   6 GB RAM, 128 GB inbu

In [ ]:
df.isna().sum()

Spec Score        0
rating            0
price             0
Name              0
img               0
tag               0
sim               0
processor         0
storage           0
battery           0
display           0
camera            0
memoryExternal    0
version           0
fm                0
dtype: int64

In [ ]:
df.duplicated().sum()

np.int64(2)

In [ ]:
df.drop_duplicates(inplace=True)

In [ ]:
df.duplicated().sum()

np.int64(0)

In [ ]:
df["rating"]=df["rating"].apply(lambda x:str(x))

In [ ]:
df["rating"]=df["rating"].apply(lambda x:x[0:3])

In [ ]:
df

Spec Score rating   price                                        Name  \
0             89    4.6   41990                         OPPO Reno 14 Pro 5G   
1             87    4.7   32990                             OPPO Reno 14 5G   
2             83    4.7   31999                                  Poco F7 5G   
3             89    4.6   49990                                Vivo X200 FE   
4             73    4.2   11999                                OPPO K13x 5G   
...          ...    ...     ...                                         ...   
1014          85    4.0   36999              Realme GT 7 Pro Racing Edition   
1015          75    4.4   29990                                iQOO Neo 11R   
1016          88    4.2  111999  Samsung Galaxy S25 Plus (12GB RAM + 512GB)   
1017          74    4.0    9699    Samsung Galaxy A14 4G (6GB RAM + 128 GB)   
1018          75    4.6   12999        Tecno Spark 30C 5G (8GB RAM + 128GB)   

                                                    img       tag  \
0     https://cdn1.smartprix.com/rx-is822PXo3-w280-h...  UPCOMING   
1     https://cdn1.smartprix.com/rx-iRGgfcGDH-w280-h...  UPCOMING   
2     https://cdn1.smartprix.com/rx-icmgBU9Q2-w280-h...  LAUNCHED   
3     https://cdn1.smartprix.com/rx-iHI7IaQgQ-w280-h...  UPCOMING   
4     https://cdn1.smartprix.com/rx-iXUulomIY-w280-h...  LAUNCHED   
...                                                 ...       ...   
1014  https://cdn1.smartprix.com/rx-iA2dbnJIw-w280-h...  UPCOMING   
1015  https://cdn1.smartprix.com/rx-iBdtueAix-w280-h...   RUMORED   
1016  https://cdn1.smartprix.com/rx-iGH33GWY9-w280-h...  LAUNCHED   
1017  https://cdn1.smartprix.com/rx-igcSDy9DV-w280-h...  LAUNCHED   
1018  https://cdn1.smartprix.com/rx-ilU0Yr0Ho-w280-h...  LAUNCHED   

                                                    sim  \
0     Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
1     Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
2     Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
3     Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
4                    Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi   
...                                                 ...   
1014  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
1015  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
1016      Dual Sim, 3G, 4G, 5G, VoLTE, Vo5G, Wi-Fi, NFC   
1017                     Dual Sim, 3G, 4G, VoLTE, Wi-Fi   
1018  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   

                                              processor  \
0         Dimensity 8450, Octa Core, 3.25 GHz Processor   
1         Dimensity 8350, Octa Core, 3.35 GHz Processor   
2      Snapdragon 8s Gen4, Octa Core, 3.2 GHz Processor   
3     Dimensity 9300 Plus, Octa Core, 3.25 GHz Proce...   
4          Dimensity 6300, Octa Core, 2.4 GHz Processor   
...                                                 ...   
1014  Snapdragon 8 Elite, Octa Core, 4.32 GHz Processor   
1015                                Octa Core Processor   
1016   Snapdragon 8Elite, Octa Core, 4.47 GHz Processor   
1017             Exynos 850, Octa Core, 2 GHz Processor   
1018       Dimensity 6300, Octa Core, 2.4 GHz Processor   

                        storage                                   battery  \
0     12 GB RAM, 256 GB inbuilt   6200 mAh Battery with 80W Fast Charging   
1      8 GB RAM, 256 GB inbuilt   6000 mAh Battery with 80W Fast Charging   
2     12 GB RAM, 256 GB inbuilt   7550 mAh Battery with 90W Fast Charging   
3     12 GB RAM, 256 GB inbuilt   6500 mAh Battery with 90W Fast Charging   
4      4 GB RAM, 128 GB inbuilt   6000 mAh Battery with 45W Fast Charging   
...                         ...                                       ...   
1014  12 GB RAM, 256 GB inbuilt  6500 mAh Battery with 120W Fast Charging   
1015  12 GB RAM, 256 GB inbuilt  6700 mAh Battery with 100W Fast Charging   
1016  12 GB RAM, 512 GB inbuilt   4900 mAh Battery with 45W Fast Charging   
1017   6 GB RAM, 128 GB inbuilt   5000 m

img (image URLs) may not be useful unless doing image analysis.

tag (LAUNCHED/UPCOMING) is categorical and may be kept, but could also be simplified.

In [ ]:
df.drop(columns=['img','tag'],inplace=True)

In [ ]:
df

Spec Score rating   price                                        Name  \
0             89    4.6   41990                         OPPO Reno 14 Pro 5G   
1             87    4.7   32990                             OPPO Reno 14 5G   
2             83    4.7   31999                                  Poco F7 5G   
3             89    4.6   49990                                Vivo X200 FE   
4             73    4.2   11999                                OPPO K13x 5G   
...          ...    ...     ...                                         ...   
1014          85    4.0   36999              Realme GT 7 Pro Racing Edition   
1015          75    4.4   29990                                iQOO Neo 11R   
1016          88    4.2  111999  Samsung Galaxy S25 Plus (12GB RAM + 512GB)   
1017          74    4.0    9699    Samsung Galaxy A14 4G (6GB RAM + 128 GB)   
1018          75    4.6   12999        Tecno Spark 30C 5G (8GB RAM + 128GB)   

                                                    sim  \
0     Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
1     Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
2     Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
3     Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
4                    Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi   
...                                                 ...   
1014  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
1015  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
1016      Dual Sim, 3G, 4G, 5G, VoLTE, Vo5G, Wi-Fi, NFC   
1017                     Dual Sim, 3G, 4G, VoLTE, Wi-Fi   
1018  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   

                                              processor  \
0         Dimensity 8450, Octa Core, 3.25 GHz Processor   
1         Dimensity 8350, Octa Core, 3.35 GHz Processor   
2      Snapdragon 8s Gen4, Octa Core, 3.2 GHz Processor   
3     Dimensity 9300 Plus, Octa Core, 3.25 GHz Proce...   
4          Dimensity 6300, Octa Core, 2.4 GHz Processor   
...                                                 ...   
1014  Snapdragon 8 Elite, Octa Core, 4.32 GHz Processor   
1015                                Octa Core Processor   
1016   Snapdragon 8Elite, Octa Core, 4.47 GHz Processor   
1017             Exynos 850, Octa Core, 2 GHz Processor   
1018       Dimensity 6300, Octa Core, 2.4 GHz Processor   

                        storage                                   battery  \
0     12 GB RAM, 256 GB inbuilt   6200 mAh Battery with 80W Fast Charging   
1      8 GB RAM, 256 GB inbuilt   6000 mAh Battery with 80W Fast Charging   
2     12 GB RAM, 256 GB inbuilt   7550 mAh Battery with 90W Fast Charging   
3     12 GB RAM, 256 GB inbuilt   6500 mAh Battery with 90W Fast Charging   
4      4 GB RAM, 128 GB inbuilt   6000 mAh Battery with 45W Fast Charging   
...                         ...                                       ...   
1014  12 GB RAM, 256 GB inbuilt  6500 mAh Battery with 120W Fast Charging   
1015  12 GB RAM, 256 GB inbuilt  6700 mAh Battery with 100W Fast Charging   
1016  12 GB RAM, 512 GB inbuilt   4900 mAh Battery with 45W Fast Charging   
1017   6 GB RAM, 128 GB inbuilt   5000 mAh Battery with 15W Fast Charging   
1018   8 GB RAM, 128 GB inbuilt   5000 mAh Battery with 18W Fast Charging   

                                                display  \
0     6.83 inches, 1272 x 2800 px, 120 Hz Display wi...   
1     6.59 inches, 1256 x 2760 px, 120 Hz Display wi...   
2     6.83 inches, 1280 x 2772 px, 120 Hz Display wi...   
3     6.31 inches, 1216 x 2640 px, 120 Hz Display wi...   
4     6.67 inches, 720 x 1604 px, 120 Hz Display wit...   
...                                                 ...   
1014  6.78 inches, 1264 x 2780 px, 120 Hz Display wi...   
1015  6.78 inches, 1260 x 2800 px, 144 Hz Display wi...   
1016  6.7 inches, 1440 x 3120 px, 120 Hz Display wit...   
1017  6.6 inches, 1080 x 2408 px Display with Water ...   
1018  6.67 inches, 720 x 1600 px, 120 Hz Display wit...   

     

Feature extraction opportunities

From battery: extract capacity (mAh) and charging power (W).

From storage: separate RAM and internal storage.

From display: extract screen size (inches), resolution, refresh rate.

From camera: extract rear and front camera megapixels.

From processor: extract chipset and clock speed.

In [ ]:
df["battery_capacity"] = df["battery"].str.extract(r'(\d+)\s*mAh')[0]+"mAh"

In [ ]:
df["charging_capacity"]=df["battery"].str.extract(r'(\d+)\s*W')[0]+"W"

In [ ]:
df.drop(columns=["battery"],inplace=True)

In [ ]:
df

Spec Score rating   price                                        Name  \
0             89    4.6   41990                         OPPO Reno 14 Pro 5G   
1             87    4.7   32990                             OPPO Reno 14 5G   
2             83    4.7   31999                                  Poco F7 5G   
3             89    4.6   49990                                Vivo X200 FE   
4             73    4.2   11999                                OPPO K13x 5G   
...          ...    ...     ...                                         ...   
1014          85    4.0   36999              Realme GT 7 Pro Racing Edition   
1015          75    4.4   29990                                iQOO Neo 11R   
1016          88    4.2  111999  Samsung Galaxy S25 Plus (12GB RAM + 512GB)   
1017          74    4.0    9699    Samsung Galaxy A14 4G (6GB RAM + 128 GB)   
1018          75    4.6   12999        Tecno Spark 30C 5G (8GB RAM + 128GB)   

                                                    sim  \
0     Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
1     Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
2     Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
3     Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
4                    Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi   
...                                                 ...   
1014  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
1015  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
1016      Dual Sim, 3G, 4G, 5G, VoLTE, Vo5G, Wi-Fi, NFC   
1017                     Dual Sim, 3G, 4G, VoLTE, Wi-Fi   
1018  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   

                                              processor  \
0         Dimensity 8450, Octa Core, 3.25 GHz Processor   
1         Dimensity 8350, Octa Core, 3.35 GHz Processor   
2      Snapdragon 8s Gen4, Octa Core, 3.2 GHz Processor   
3     Dimensity 9300 Plus, Octa Core, 3.25 GHz Proce...   
4          Dimensity 6300, Octa Core, 2.4 GHz Processor   
...                                                 ...   
1014  Snapdragon 8 Elite, Octa Core, 4.32 GHz Processor   
1015                                Octa Core Processor   
1016   Snapdragon 8Elite, Octa Core, 4.47 GHz Processor   
1017             Exynos 850, Octa Core, 2 GHz Processor   
1018       Dimensity 6300, Octa Core, 2.4 GHz Processor   

                        storage  \
0     12 GB RAM, 256 GB inbuilt   
1      8 GB RAM, 256 GB inbuilt   
2     12 GB RAM, 256 GB inbuilt   
3     12 GB RAM, 256 GB inbuilt   
4      4 GB RAM, 128 GB inbuilt   
...                         ...   
1014  12 GB RAM, 256 GB inbuilt   
1015  12 GB RAM, 256 GB inbuilt   
1016  12 GB RAM, 512 GB inbuilt   
1017   6 GB RAM, 128 GB inbuilt   
1018   8 GB RAM, 128 GB inbuilt   

                                                display  \
0     6.83 inches, 1272 x 2800 px, 120 Hz Display wi...   
1     6.59 inches, 1256 x 2760 px, 120 Hz Display wi...   
2     6.83 inches, 1280 x 2772 px, 120 Hz Display wi...   
3     6.31 inches, 1216 x 2640 px, 120 Hz Display wi...   
4     6.67 inches, 720 x 1604 px, 120 Hz Display wit...   
...                                                 ...   
1014  6.78 inches, 1264 x 2780 px, 120 Hz Display wi...   
1015  6.78 inches, 1260 x 2800 px, 144 Hz Display wi...   
1016  6.7 inches, 1440 x 3120 px, 120 Hz Display wit...   
1017  6.6 inches, 1080 x 2408 px Display with Water ...   
1018  6.67 inches, 720 x 1600 px, 120 Hz Display wit...   

                                                 camera  \
0     50 MP + 50 MP + 50 MP Triple Rear & 50 MP Fron...   
1     50 MP + 50 MP + 8 MP Triple Rear & 50 MP Front...   
2           50 MP + 8 MP Dual Rear & 20 MP Front Camera   
3     50 MP + 50 MP + 8 MP Triple Rear & 50 MP Front...   
4            50 MP + 2 MP Dual Rear & 8 MP Front Camera   
...                                                 ...   
1014        50 MP + 8 MP Dual Rear & 16 MP Front Camera   
1015       50 MP + 50 MP Dual Rear & 

In [ ]:
df["ram"]=df["storage"].str.extract(r'(\d+)\s*GB RAM')[0]+"GB RAM"
df["internal_storage"]=df["storage"].str.extract(r'(\d+)\s*GB')[0]+"GB"

In [ ]:
#From display: extract screen size (inches), resolution, refresh rate.
df["screen_size"]=df["display"].str.extract(r'(\d+)\s*inches')[0]+'inches'
df["resolution"]=df["display"].str.extract(r'(\d+\s*×\s*\d+\s*px|\d+\s*x\s*\d+)')[0]+'px'
df["refresh_rate"]=df["display"].str.extract(r'(\d+)\s*Hz')[0]+'Hz'

In [ ]:
#From camera: extract rear and front camera megapixels.
df['rear_camera']=df['camera'].str.extract(r'^(.*?)&')
df["front_camera"]=df['camera'].str.extract(r'&\s*(.*)')

In [ ]:
#From processor: extract chipset and clock speed.
df["chipset"] = df["processor"].str.extract(r'^(.*?),')
df["clock_speed"] = df["processor"].str.extract(r'(\d+\.?\d*\s*GHz)')

In [ ]:
df

Spec Score rating   price                                        Name  \
0             89    4.6   41990                         OPPO Reno 14 Pro 5G   
1             87    4.7   32990                             OPPO Reno 14 5G   
2             83    4.7   31999                                  Poco F7 5G   
3             89    4.6   49990                                Vivo X200 FE   
4             73    4.2   11999                                OPPO K13x 5G   
...          ...    ...     ...                                         ...   
1014          85    4.0   36999              Realme GT 7 Pro Racing Edition   
1015          75    4.4   29990                                iQOO Neo 11R   
1016          88    4.2  111999  Samsung Galaxy S25 Plus (12GB RAM + 512GB)   
1017          74    4.0    9699    Samsung Galaxy A14 4G (6GB RAM + 128 GB)   
1018          75    4.6   12999        Tecno Spark 30C 5G (8GB RAM + 128GB)   

                                                    sim  \
0     Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
1     Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
2     Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
3     Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
4                    Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi   
...                                                 ...   
1014  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
1015  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
1016      Dual Sim, 3G, 4G, 5G, VoLTE, Vo5G, Wi-Fi, NFC   
1017                     Dual Sim, 3G, 4G, VoLTE, Wi-Fi   
1018  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   

                                              processor  \
0         Dimensity 8450, Octa Core, 3.25 GHz Processor   
1         Dimensity 8350, Octa Core, 3.35 GHz Processor   
2      Snapdragon 8s Gen4, Octa Core, 3.2 GHz Processor   
3     Dimensity 9300 Plus, Octa Core, 3.25 GHz Proce...   
4          Dimensity 6300, Octa Core, 2.4 GHz Processor   
...                                                 ...   
1014  Snapdragon 8 Elite, Octa Core, 4.32 GHz Processor   
1015                                Octa Core Processor   
1016   Snapdragon 8Elite, Octa Core, 4.47 GHz Processor   
1017             Exynos 850, Octa Core, 2 GHz Processor   
1018       Dimensity 6300, Octa Core, 2.4 GHz Processor   

                        storage  \
0     12 GB RAM, 256 GB inbuilt   
1      8 GB RAM, 256 GB inbuilt   
2     12 GB RAM, 256 GB inbuilt   
3     12 GB RAM, 256 GB inbuilt   
4      4 GB RAM, 128 GB inbuilt   
...                         ...   
1014  12 GB RAM, 256 GB inbuilt   
1015  12 GB RAM, 256 GB inbuilt   
1016  12 GB RAM, 512 GB inbuilt   
1017   6 GB RAM, 128 GB inbuilt   
1018   8 GB RAM, 128 GB inbuilt   

                                                display  \
0     6.83 inches, 1272 x 2800 px, 120 Hz Display wi...   
1     6.59 inches, 1256 x 2760 px, 120 Hz Display wi...   
2     6.83 inches, 1280 x 2772 px, 120 Hz Display wi...   
3     6.31 inches, 1216 x 2640 px, 120 Hz Display wi...   
4     6.67 inches, 720 x 1604 px, 120 Hz Display wit...   
...                                                 ...   
1014  6.78 inches, 1264 x 2780 px, 120 Hz Display wi...   
1015  6.78 inches, 1260 x 2800 px, 144 Hz Display wi...   
1016  6.7 inches, 1440 x 3120 px, 120 Hz Display wit...   
1017  6.6 inches, 1080 x 2408 px Display with Water ...   
1018  6.67 inches, 720 x 1600 px, 120 Hz Display wit...   

                                                 camera  \
0     50 MP + 50 MP + 50 MP Triple Rear & 50 MP Fron...   
1     50 MP + 50 MP + 8 MP Triple Rear & 50 MP Front...   
2           50 MP + 8 MP Dual Rear & 20 MP Front Camera   
3     50 MP + 50 MP + 8 MP Triple Rear & 50 MP Front...   
4            50 MP + 2 MP Dual Rear & 8 MP Front Camera   
...                                                 ...   
1014        50 MP + 8 MP Dual Rear & 16 MP Front Camera   
1015       50 MP + 50 MP Dual Rear & 

In [ ]:
df.columns

Index(['Spec Score', 'rating', 'price', 'Name', 'sim', 'processor', 'storage',
       'display', 'camera', 'memoryExternal', 'version', 'fm',
       'battery_capacity', 'charging_capacity', 'ram', 'internal_storage',
       'screen_size', 'resolution', 'refresh_rate', 'rear_camera',
       'front_camera', 'chipset', 'clock_speed'],
      dtype='object')

In [ ]:
df.drop(columns=['battery_capacity','storage','camera','processor','display'],inplace=True)

In [ ]:
df

Spec Score rating   price                                        Name  \
0             89    4.6   41990                         OPPO Reno 14 Pro 5G   
1             87    4.7   32990                             OPPO Reno 14 5G   
2             83    4.7   31999                                  Poco F7 5G   
3             89    4.6   49990                                Vivo X200 FE   
4             73    4.2   11999                                OPPO K13x 5G   
...          ...    ...     ...                                         ...   
1014          85    4.0   36999              Realme GT 7 Pro Racing Edition   
1015          75    4.4   29990                                iQOO Neo 11R   
1016          88    4.2  111999  Samsung Galaxy S25 Plus (12GB RAM + 512GB)   
1017          74    4.0    9699    Samsung Galaxy A14 4G (6GB RAM + 128 GB)   
1018          75    4.6   12999        Tecno Spark 30C 5G (8GB RAM + 128GB)   

                                                    sim  \
0     Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
1     Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
2     Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
3     Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
4                    Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi   
...                                                 ...   
1014  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
1015  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
1016      Dual Sim, 3G, 4G, 5G, VoLTE, Vo5G, Wi-Fi, NFC   
1017                     Dual Sim, 3G, 4G, VoLTE, Wi-Fi   
1018  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   

                        memoryExternal      version           fm  \
0            Memory Card Not Supported  Android v15  No FM Radio   
1            Memory Card Not Supported  Android v15  No FM Radio   
2            Memory Card Not Supported  Android v15  No FM Radio   
3            Memory Card Not Supported  Android v15  No FM Radio   
4     Memory Card Supported, upto 2 TB  Android v15  No FM Radio   
...                                ...          ...          ...   
1014         Memory Card Not Supported  Android v15  No FM Radio   
1015         Memory Card Not Supported  Android v16  No FM Radio   
1016         Memory Card Not Supported  Android v15  No FM Radio   
1017  Memory Card Supported, upto 1 TB  Android v13  No FM Radio   
1018  Memory Card Supported, upto 1 TB  Android v14  No FM Radio   

     charging_capacity       ram internal_storage screen_size     resolution  \
0                  80W  12GB RAM             12GB    83inches  1272 x 2800px   
1                  80W   8GB RAM              8GB    59inches  1256 x 2760px   
2                  90W  12GB RAM             12GB    83inches  1280 x 2772px   
3                  90W  12GB RAM             12GB    31inches  1216 x 2640px   
4                  45W   4GB RAM              4GB    67inches   720 x 1604px   
...                ...       ...              ...         ...            ...   
1014              120W  12GB RAM             12GB    78inches  1264 x 2780px   
1015              100W  12GB RAM             12GB    78inches  1260 x 2800px   
1016               45W  12GB RAM             12GB     7inches  1440 x 3120px   
1017               15W   6GB RAM              6GB     6inches  1080 x 2408px   
1018               18W   8GB RAM              8GB    67inches   720 x 1600px   

     refresh_rate                         rear_camera        front_camera  \
0           120Hz  50 MP + 50 MP + 50 MP Triple Rear   50 MP Front Camera   
1           120Hz   50 MP + 50 MP + 8 MP Triple Rear   50 MP Front Camera   
2           120Hz             50 MP + 8 MP Dual Rear   20 MP Front Camera   
3           120Hz   50 MP + 50 MP + 8 MP Triple Rear   50 MP Front Camera   
4           120Hz             50 MP + 2 MP Dual Rear    8 MP Front Camera   
...           ...                                 ...                 ...   
1014        120Hz             5

In [ ]:
#We are going to change the position of the fm column and inserting it to the last of the tabel

col = df.pop("fm")
df.insert(len(df.columns)-1, "fm", col)

In [ ]:
df

Spec Score rating   price                                        Name  \
0             89    4.6   41990                         OPPO Reno 14 Pro 5G   
1             87    4.7   32990                             OPPO Reno 14 5G   
2             83    4.7   31999                                  Poco F7 5G   
3             89    4.6   49990                                Vivo X200 FE   
4             73    4.2   11999                                OPPO K13x 5G   
...          ...    ...     ...                                         ...   
1014          85    4.0   36999              Realme GT 7 Pro Racing Edition   
1015          75    4.4   29990                                iQOO Neo 11R   
1016          88    4.2  111999  Samsung Galaxy S25 Plus (12GB RAM + 512GB)   
1017          74    4.0    9699    Samsung Galaxy A14 4G (6GB RAM + 128 GB)   
1018          75    4.6   12999        Tecno Spark 30C 5G (8GB RAM + 128GB)   

                                                    sim  \
0     Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
1     Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
2     Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
3     Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
4                    Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi   
...                                                 ...   
1014  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
1015  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
1016      Dual Sim, 3G, 4G, 5G, VoLTE, Vo5G, Wi-Fi, NFC   
1017                     Dual Sim, 3G, 4G, VoLTE, Wi-Fi   
1018  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   

                        memoryExternal      version charging_capacity  \
0            Memory Card Not Supported  Android v15               80W   
1            Memory Card Not Supported  Android v15               80W   
2            Memory Card Not Supported  Android v15               90W   
3            Memory Card Not Supported  Android v15               90W   
4     Memory Card Supported, upto 2 TB  Android v15               45W   
...                                ...          ...               ...   
1014         Memory Card Not Supported  Android v15              120W   
1015         Memory Card Not Supported  Android v16              100W   
1016         Memory Card Not Supported  Android v15               45W   
1017  Memory Card Supported, upto 1 TB  Android v13               15W   
1018  Memory Card Supported, upto 1 TB  Android v14               18W   

           ram internal_storage screen_size     resolution refresh_rate  \
0     12GB RAM             12GB    83inches  1272 x 2800px        120Hz   
1      8GB RAM              8GB    59inches  1256 x 2760px        120Hz   
2     12GB RAM             12GB    83inches  1280 x 2772px        120Hz   
3     12GB RAM             12GB    31inches  1216 x 2640px        120Hz   
4      4GB RAM              4GB    67inches   720 x 1604px        120Hz   
...        ...              ...         ...            ...          ...   
1014  12GB RAM             12GB    78inches  1264 x 2780px        120Hz   
1015  12GB RAM             12GB    78inches  1260 x 2800px        144Hz   
1016  12GB RAM             12GB     7inches  1440 x 3120px        120Hz   
1017   6GB RAM              6GB     6inches  1080 x 2408px          NaN   
1018   8GB RAM              8GB    67inches   720 x 1600px        120Hz   

                             rear_camera        front_camera  \
0     50 MP + 50 MP + 50 MP Triple Rear   50 MP Front Camera   
1      50 MP + 50 MP + 8 MP Triple Rear   50 MP Front Camera   
2                50 MP + 8 MP Dual Rear   20 MP Front Camera   
3      50 MP + 50 MP + 8 MP Triple Rear   50 MP Front Camera   
4                50 MP + 2 MP Dual Rear    8 MP Front Camera   
...                                  ...                 ...   
1014             50 MP + 8 MP Dual Rear   16 MP Front Camera   
1015            50 MP + 50 MP Dual Rear   32 MP Front Came

In [ ]:
df.to_csv("Mobile_cleaned",index=False)